extensions/message-tags.md

```diff
@@ -44,8 +44,8 @@ The message pseudo-BNF, as defined in [RFC 1459, section 2.3.1][rfc1459] is exte
     <tag>           ::= <key> ['=' <escaped_value>]
     <key>           ::= [ <client_prefix> ] [ <vendor> '/' ] <key_name>
     <client_prefix> ::= '+'
-    <key_name>      ::= <sequence of letters, digits, hyphens ('-')>
-    <escaped_value> ::= <sequence of any characters except NUL, CR, LF, semicolon (`;`) and SPACE>
+    <key_name>      ::= <non-empty sequence of ascii letters, digits, hyphens ('-')>
+    <escaped_value> ::= <sequence of zero or more utf8 characters except NUL, CR, LF, semicolon (`;`) and SPACE>
```

Comment:
This retroactively breaks networks which use non-utf8 charset

Reply:
Which existing tags allow non utf8?

Reply:
In the value? Any non-standard one.

But I was more refering to the concept of requiring UTF-8. E.g. in ZNC charset is applied to the whole line, before parsing of tags even begins. I hope you don't want UTF-8 to be encoded inside some other encoding.

Reply:
How is znc going to handle client sent utf8 in future?

Reply:
Also, we don’t need to care about non standard implementations.

Reply:
IRC is defined in terms of octets, not UTF-8 text. If we are to change that, that should be done explicitly, in a separate spec, not retroactively baked into some obscure line in one of unrelated specs.

> How is znc going to handle client sent utf8 in future?

There are settings for that.

Reply:
I am focused on new parts of the protocol here, not changing things retroactively. I want to ensure everyone can assume a sane encoding for new features, so we can avoid some of the short sightedness of the past. This seemed like a good place to enforce this for new tags. What would be a more appropriate way of doing this?

Reply:
> There are settings for that.

Does this mean the default would already be “broken” behaviour as you call it? I’m not sure how this changes anything in that case.

Reply:
> What would be a more appropriate way of doing this?

Instead of changing things subtly and retroactively, change them explicitly. I generally agree with usage of UTF-8 everywhere. But need to write a separate spec which defines UTF-8 as the encoding, which discusses how to enable this (cap? implicitly?), and how it should interact with other protocol features, with existing software out there.

> Does this mean the default would already be “broken” behaviour as you call it?

Define "broken".

> I’m not sure how this changes anything in that case.

Here's an example.

Client <--- UTF-8 ---> ZNC <--- KOI8-R ---> Server

If client sends a message, ZNC interprets it as UTF-8, but sends it to server as KOI8-R. If server sends a message, ZNC decodes it from KOI8-R, and sends to client as UTF-8.

Now imagine that the message had some tags, and the value of one of characters in some tag is outside of ASCII range. What exactly is supposed to happen here?

Reply:
Broken was your word, I’m not sure how you’re defining it.

With your example, what would you do in znc right now if that happens?

Reply:
Now the whole line gets reencoded, including tags. So by changing the requirement for tags to be utf-8, you're retroactively breaking existing software, which suddenly wouldn't be conformant anymore.

Reply:
The KOI8-R "server" you're talking about is actually just an encoding unaware relay. The KOI8-R data you're dealing with is in message bodies that users on that server have decided to use amongst each other. The clients of those users could already start sending utf8 data in tags while still allowing KOI8-R in messages. The server could also add utf8 tag data of its own if it wanted. Your whole line encoding might already not deal with that very well so it's already a "broken" situation you will likely need to address. And if you can assume tags are always utf8, this is actually much easier for you to address.

All I'm proposing is that defined, specified new tag values MUST be utf8. This is designed to make things easier. I don't think the existing tags we have (server-time, account-tag, batch) are at much risk of containing non-utf8 data.

Reply:
> The KOI8-R "server" you're talking about is actually just an encoding unaware relay. 

Unfortunately, it's not. Server actually has several ports, and uses different encoding on each. So client connected on port 6667 and client connected on port 6668 use different encodings, while still talking to each other.

> The server could also add utf8 tag data of its own if it wanted

So where is the split between utf8 and the other encoding? End of tags? Again, this needs a separate spec and a separate discussion.

Reply:
Ok so in that case, your russian re-encoding server will have to deal with this, not you. And if it's in the spec, that's doable. I think we can agree this is a rather niche case, isn't it?

Why is the tags spec not a good place to specify the encoding of tags?

Reply:
I've removed the "utf8" bit from this PR for now and will move it its own PR.